File: solutions/cpp/BST_to_sorted_doubly_list.cc
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.

#include <cassert>
#include <iostream>
#include <limits>
#include <memory>

#include "./BST_prototype_shared_ptr.h"

using std::cout;
using std::endl;
using std::make_shared;
using std::numeric_limits;
using std::pair;
using std::shared_ptr;

pair<shared_ptr<BSTNode<int>>, shared_ptr<BSTNode<int>>>
    BSTToDoublyLinkedListHelper(const shared_ptr<BSTNode<int>>&);

// @include
shared_ptr<BSTNode<int>> BSTToDoublyLinkedList(
    const shared_ptr<BSTNode<int>>& tree) {
  return BSTToDoublyLinkedListHelper(tree).first;
}

// Transforms a BST into a sorted doubly linked list in-place,
// and return the head and tail of the list as a pair.
pair<shared_ptr<BSTNode<int>>, shared_ptr<BSTNode<int>>>
    BSTToDoublyLinkedListHelper(const shared_ptr<BSTNode<int>>& tree) {
  // Empty subtree.
  if (!tree) {
    return {nullptr, nullptr};
  }

  // Recursively builds the list from left and right subtrees.
  auto left = BSTToDoublyLinkedListHelper(tree->left);
  auto right = BSTToDoublyLinkedListHelper(tree->right);

  // Appends tree to the list from left subtree.
  if (left.second) {
    left.second->right = tree;
  }
  tree->left = left.second;

  // Appends the list from right subtree to tree.
  tree->right = right.first;
  if (right.first) {
    right.first->left = tree;
  }

  return {left.first ? left.first : tree, right.second ? right.second : tree};
}
// @exclude

int main(int argc, char* argv[]) {
  //    3
  //  2   5
  // 1   4 6
  auto root = make_shared<BSTNode<int>>(BSTNode<int>{3});
  root->left = make_shared<BSTNode<int>>(BSTNode<int>{2});
  root->left->left = make_shared<BSTNode<int>>(BSTNode<int>{1});
  root->right = make_shared<BSTNode<int>>(BSTNode<int>{5});
  root->right->left = make_shared<BSTNode<int>>(BSTNode<int>{4});
  root->right->right = make_shared<BSTNode<int>>(BSTNode<int>{6});
  shared_ptr<BSTNode<int>> L = BSTToDoublyLinkedList(root);
  shared_ptr<BSTNode<int>> curr = L;
  int pre = numeric_limits<int>::min();
  do {
    assert(pre <= curr->data);
    cout << curr->data << endl;
    pre = curr->data;
    curr = curr->right;
  } while (curr);
  return 0;
}

File: static/solutions/java/src/main/java/com/epi/BSTToSortedDoublyList.java
package com.epi;

import com.epi.BinarySearchTreePrototypeTemplate.BSTNode;

public class BSTToSortedDoublyList {
  // @include
  private static class HeadAndTail {
    public BSTNode<Integer> head;
    public BSTNode<Integer> tail;

    public HeadAndTail(BSTNode<Integer> head, BSTNode<Integer> tail) {
      this.head = head;
      this.tail = tail;
    }
  }

  public static BSTNode<Integer> bstToDoublyLinkedList(BSTNode<Integer> tree) {
    return bstToDoublyLinkedListHelper(tree).head;
  }

  // Transforms a BST into a sorted doubly linked list in-place, and return the
  // head and tail of the list.
  private static HeadAndTail bstToDoublyLinkedListHelper(
      BSTNode<Integer> tree) {
    // Empty subtree.
    if (tree == null) {
      return new HeadAndTail(null, null);
    }

    // Recursively build the list from left and right subtrees.
    HeadAndTail left = bstToDoublyLinkedListHelper(tree.left);
    HeadAndTail right = bstToDoublyLinkedListHelper(tree.right);

    // Append tree to the list from left subtree.
    if (left.tail != null) {
      left.tail.right = tree;
    }
    tree.left = left.tail;

    // Append the list from right subtree to tree.
    tree.right = right.head;
    if (right.head != null) {
      right.head.left = tree;
    }

    return new HeadAndTail(left.head != null ? left.head : tree,
                           right.tail != null ? right.tail : tree);
  }
  // @exclude

  public static void main(String[] args) {
    // 3
    // 2 5
    // 1 4 6
    BSTNode<Integer> tree = new BSTNode<>(3);
    tree.left = new BSTNode<>(2);
    tree.left.left = new BSTNode<>(1);
    tree.right = new BSTNode<>(5);
    tree.right.left = new BSTNode<>(4);
    tree.right.right = new BSTNode<>(6);
    BSTNode<Integer> L = bstToDoublyLinkedList(tree);
    BSTNode<Integer> curr = L;
    int pre = Integer.MIN_VALUE;
    do {
      assert(pre <= curr.data);
      System.out.println(curr.data);
      pre = curr.data;
      curr = curr.right;
    } while (curr != null);
  }
}
